### ISSUE TITLE
In searchable ActiveModel integration, add a proxy object to every result item, pointing to the model instance

### ISSUE THREAD

**Issue by karmi** [karmi/retire#12]
Due to the problems with ActiveRecord models with associations (see the original issue karmi/tire#10), add a proxy object to every result item allowing the access to the underlying model instance.

This way, we would keep the performance of ElasticSearch, for most cases, and allow to access the original model instance (and its methods), where needed. So, instead of `result.comments.first.my_complicated_method`, you would write `result.object.comments.first.my_complicated_method`.

**Comment by floere** [comment]
How about "model", since it is always an ActiveModel?

**Comment by karmi** [comment]
Yes, I think `model` is better then `object`... But none feels 100% right to me...

**Comment by floere** [comment]
Yes. I suppose that is because long method chains always are problematic. Writing tests for `result.object.comments.first.my_complicated_method` will not be fun.

**Comment by pfeiffer** [comment]
Would a solution be to let the developer specify when to use 'real' records and when to use data returned by ElasticSearch on a case by case basis? The developer could choose full model availability over performance where needed.

Preferably there wouldn't be a difference between handling results returned from `Model.search` and eg. `Model.all`

**Comment by karmi** [comment]
This issue is a tricky one, and there are no easy shortcuts.

Of course, the ideal solution from the user's point of view would be to return full model instances, so there's no difference between `Model.search` and `Model.all`, as you write.

There are, however, lots of problems.

1/ Of course, we can always just load the records from the underlying datastore, by their IDs. But, as stated in karmi/tire#10, this would seriously damage the search performance. ElasticSearch returns hits for reasonable datasets or with enough firepower (RAM+nodes) in the tens of milliseconds magnitude. The roundtrips to the db could make this very costly -- and let's remember, the usage in Rails, with the additional overhead of HTML construction, etc -- is not the only use-case. There are use-cases where the raw speed does matter a lot. (Think Ajax autocompleter, for instance.)

2/ `ActiveRecord`, by default, does not preserve the order of IDs. So when you do a `MyModel.find [1, 2, 3]`, the order really isn't guaranteed. This is really bad, because the search engine tries very hard to actually _do sort_ the results based on relevancy. Firing 30-50 queries against SQL or CouchDB really isn't that fast.

3/ This all is still on the technical level, solvable probably. But let's have a user who indexes her articles like this:

``` ruby
    class Article < ActiveRecord::Base
      # ...
      def to_indexed_json
        { :id      => id,
          :title   => title,
          :content => content,
          :categories => comments.map(&:name),

          # >>> NOTICE THIS ONE >>>
          :length  = content.size
          # <<<<<<<<<<<<<<<<<<<<<<<
        }.to_json
      end
    end
```

Why would she index the length of the article like this? Well, because she wants to run a [statistical facet](http://www.elasticsearch.org/guide/reference/api/search/facets/statistical-facet.html) against the length, of course, retrieving min/max/avg/etc calculations!

Now what? Does she have to implement an `#length` accessor methods just for the sake of _Tire_? That does not make any sense. She could be perfectly happy with displaying `article.title` and `results.facets['stats']['mean']`.

If, and only if, she would like to display something not present in the JSON data from ElasticSearch, such as `Article#collaborators`, well she would just write `article.<PROXY>.collaborators`.

Notice I still don't know what's the best name for the proxy object.

Maybe `load`? Since of course, the object would have to be lazy loaded upon invocation, so there would be _no_ roundtrips to database when you'd be satisfied with the JSON data returned from ElasticSearch.

Does it make sense like this?

I keep thinking this is the best balance between ElasticSearch performance, results usability and deep integration into some complicated relationships between models.

P.S.
Many thanks to my man @kubicek for the idea and discussion.

**Comment by pfeiffer** [comment]
I agree - there are no easy shortcuts.

If we go with the proxy object, I at least think the developer should be given the choice to eager-load the records or we might hit a N+1 situation which could impact performance even more. Since the order of the records returned by `Model.find [1, 2, 3]` isn't preserved it's important to order the returned records according to the IDs returned by ElasticSearch. In most use cases I can think of this should be reasonable, since the number of records that we need to sort is fairly small (30-100).

One thing that could help performance in this case, is returning a result set proxy that only does the sorting and retrieval of records when needed, eg. when accessing the collection - kinda like lazy relations in Arel. This way we can do a `results = User.search(..); results.facets` without hitting the underlying datastore, and only hit the datastore when we do `results.each ...`

**Comment by floere** [comment]
@karmi: Referring to 2) I had the same problem(s) in the picky-client. It is not so bad actually, performance-wise, to reorder them. Annoying, yes, but not bad at all. See https://github.com/floere/picky/blob/master/client/lib/picky-client/convenience.rb#L64-75.

**Comment by karmi** [comment]
@floere: Yes, it'd be trivial to sort the results based on sort order from ES results. And yes, it would be faster then firing queries. However, this is the easisest one of those problems.

@pfeiffer: I would definitely aim for _not hitting the datastore_, unless absolutely necessary and needed. It just doesn't make any sense to sacrifice speed. Speed is the currency of web applications.

And yes, it would certainly hit N+1 problems. However, since the results are paged in a typical Rails scenario, the N would be kept rather low, I think. Unless we add some options like `:include` to the search method, which is then passed to `find`, there is no ultimate solution for this. But maybe I'm missing something, I really haven't extensively used an SQL-based datastore for more then two years :)

It's a question of balance here. I'll bite into this next week, with some realistic use case (searching support tickets for a hosting company), and we'll see how it goes. While I think painless integration into Rails app is the absolute requirement, let's not get distracted by `ActiveRecord` complexities too much. In couple of years, the default storage for Rails _will_ change to a document-oriented storage , that's for sure.

**Comment by jkraemer** [comment]
With acts_as_ferret I built a wrapper around the hash returned from the search engine for each result which would only load the underlying model object when an attribute not being in the hash was queried.

Looking back I think that's overkill, because as a  developer you usually know if you're going to use data from the DB on your results page or not. So an option telling Tire to load (and order according to score) all result records from DB when running the search should be enough imho. However it would still be nice if one could use article.some_attribute instead of article.model.some_attribute when querying an attribute of the prefetched AR model...

**Comment by karmi** [comment]
@jkraemer: Thanks for the feedback, Jens. There's a great chance you're right, and everything else than simply loading the records from the database is probably an overkill.

As evident from the closing commit, I thought that giving trying this "proxy" approach is worth giving a try. I did it so it preserves all the elasticity of simple JSON serializations, and allows loading the records from the database, complete with support for stuff like `Article.find 1, :include => 'comments'`.

As stated in tire#10, I already had the loading from storage implemented in karmi/tire@5d89dd1. It quite obviously worked, but I wasn't very happy with the semantics, though.

I'm trying hard for the _Tire_ API to be straightforward, minimal and generally out of your way. All the different options must be documented somewhere, etc. That said, I can well imagine that I'll implement some option for the `search` methods to eagerly adn simply load records from the datastore. I can imagine, there are situations, where that is preferrable.

The premise right now is that it's generally preferrable to have easy access for properties stored in _ElasticSearch_ with the dot notation (`article.comments.first.title`), and to be able to load the "real" model (`article.load(:include => 'comments')`), when needed.

I'd be very happy if you could try this in your apps, the only thing needed is to load the gem from Github:

```
gem "tire", :git => "git://github.com/karmi/tire.git", :branch => "activerecord"
```

One final thought: I'm doing all this because my intuition tells me to keep search results as far from the database as possible, and it's something we have agreed upon with @kimchy. _ElasticSearch_ does an awesome job in storing _and_ returning data, and I don't want to hamper it by issues coming from database roundtrips. In fact, when the search results does not contain all the data you need to display them, you _may_ be doing something wrong.

**Comment by karmi** [comment]
OK, it seems like I may have found some balance in this.

By default, records are loaded from _ElasticSearch_ and instantiated as `Results::Item` instances, as described in this issue.

When needed, you can pass and option like `:load => true` to search methods, to load records directly from database, based on IDs returned from _ElasticSearch_ (see issue karmi/tire#61):

``` ruby
    # Will call `Article.search [1, 2, 3]`
    Article.search 'love', :load => true
```

You can pass any option eg. for ActiveRecord instead of true, of course:

``` ruby
    # Will call `Article.search [1, 2, 3], :include => 'comments'`
    Article.search :load => { :include => 'comments' } do
      query { string 'love' }
    end
```

I'm all ears for any feedback on this.
